File: backend/models/story_session.py
"""
Canonical Story Session State — 2026-05-16 P0 (Stability Sprint Session 2)
============================================================================

This module defines the SINGLE SOURCE OF TRUTH for a user's story editing
session. It replaces the ad-hoc collection of `useState`/`localStorage`/
`async-callback` ownership that previously caused race conditions, stale
overwrites and phantom UI.

Design pillars (founder-mandated, in order):

  1.  ONE canonical state model            (`StorySessionState`)
  2.  Explicit lifecycle transitions       (`Lifecycle` enum + table)
  3.  Illegal transition guards            (`is_legal_transition`)
  4.  Draft version incrementing           (`version` field, monotonic)
  5.  Stale write rejection                (`StaleWriteError`)
  6.  Deterministic hydration              (one read endpoint, no merging)
  7.  Single source of truth               (every write through the service)
  8.  Immutable updates                    (model is frozen / copy_with patches)
  9.  Strict ownership by draft_id/job_id  (composite key on every op)
 10.  Regression coverage first-class      (every transition + error tested)

Scope discipline
----------------
This file is pure data + state machine + error types. ZERO IO. ZERO routes.
The route layer (`routes/drafts.py`) and service layer
(`services/story_session_service.py`) compose this primitive — they never
mutate fields directly.

Backward compatibility
----------------------
Older draft documents (pre-Session-2) are missing the `version` and
`lifecycle` fields. The service backfills them on first read so existing
sessions don't have to be migrated by a batch job.
"""
from __future__ import annotations

import enum
from datetime import datetime, timezone
from typing import Any, Dict, FrozenSet, Optional, Tuple
from pydantic import BaseModel, Field, ConfigDict


# ════════════════════════════════════════════════════════════════════════════
# Lifecycle — explicit, finite state machine
# ════════════════════════════════════════════════════════════════════════════

class Lifecycle(str, enum.Enum):
    """All legal lifecycle states for a story session.

    The string values are stable wire-format identifiers. NEVER rename them
    without a migration — they are persisted on every draft document and
    cited in error envelopes consumed by the frontend.
    """
    IDLE = "IDLE"                              # fresh blank draft, no edits yet
    EDITING = "EDITING"                        # user is actively typing
    AUTOSAVING = "AUTOSAVING"                  # autosave roundtrip in flight
    READY_TO_GENERATE = "READY_TO_GENERATE"    # has content, passed validation
    GENERATING = "GENERATING"                  # pipeline job attached & running
    READY = "READY"                            # generation completed successfully
    FAILED = "FAILED"                          # generation failed (recoverable)
    ARCHIVED = "ARCHIVED"                      # soft-deleted (Start Fresh)


# All edges in the state graph. Frozen at module load — any change requires
# a documented migration. The shape is `from_state -> {allowed next states}`.
#
# Reading the graph:
#   • IDLE is the only entry state for a fresh draft.
#   • ARCHIVED is the only terminal state — once archived, the draft is
#     immutable from the user's perspective (still admin-recoverable).
#   • READY can flow back to EDITING ONLY by way of a fresh draft created
#     elsewhere — same draft_id cannot be "un-readied" without an explicit
#     architectural decision. Today we forbid READY -> EDITING to keep the
#     contract clean; remix-into-same-draft would be a future migration.
_LEGAL_TRANSITIONS: Dict[Lifecycle, FrozenSet[Lifecycle]] = {
    Lifecycle.IDLE: frozenset({
        Lifecycle.EDITING,
        Lifecycle.ARCHIVED,
    }),
    Lifecycle.EDITING: frozenset({
        Lifecycle.AUTOSAVING,
        Lifecycle.READY_TO_GENERATE,
        Lifecycle.ARCHIVED,
    }),
    Lifecycle.AUTOSAVING: frozenset({
        Lifecycle.EDITING,             # user kept typing while save was in flight
        Lifecycle.READY_TO_GENERATE,   # autosave finished + content is valid
        Lifecycle.ARCHIVED,
    }),
    Lifecycle.READY_TO_GENERATE: frozenset({
        Lifecycle.EDITING,             # user edited more before clicking Generate
        Lifecycle.GENERATING,
        Lifecycle.ARCHIVED,
    }),
    Lifecycle.GENERATING: frozenset({
        Lifecycle.READY,
        Lifecycle.FAILED,
        # NOTE: ARCHIVED from GENERATING is intentionally forbidden — the
        # in-flight pipeline job owns the document until it terminates. The
        # frontend's "Start Fresh" while generating must first wait for the
        # job to finish (or be cancelled via the pipeline's own cancel API,
        # which transitions to FAILED).
    }),
    Lifecycle.READY: frozenset({
        Lifecycle.ARCHIVED,
    }),
    Lifecycle.FAILED: frozenset({
        Lifecycle.EDITING,             # user can edit + retry
        Lifecycle.ARCHIVED,
    }),
    Lifecycle.ARCHIVED: frozenset(),   # terminal
}


def is_legal_transition(prev: Lifecycle, nxt: Lifecycle) -> bool:
    """Return True iff `prev -> nxt` is allowed by the canonical state graph."""
    if prev == nxt:
        # Re-asserting the same state is always legal (idempotent retries).
        return True
    return nxt in _LEGAL_TRANSITIONS.get(prev, frozenset())


def legal_next_states(prev: Lifecycle) -> FrozenSet[Lifecycle]:
    """Return the frozen set of legal next states from `prev`."""
    return _LEGAL_TRANSITIONS.get(prev, frozenset())


# Legacy `status` field is still persisted alongside `lifecycle` so older
# endpoints (`/current`, `/recent`) keep working without a schema migration.
# This mapping is the SINGLE PLACE that knows how the two fields relate.
_LIFECYCLE_TO_LEGACY_STATUS: Dict[Lifecycle, str] = {
    Lifecycle.IDLE: "draft",
    Lifecycle.EDITING: "draft",
    Lifecycle.AUTOSAVING: "draft",
    Lifecycle.READY_TO_GENERATE: "draft",
    Lifecycle.GENERATING: "processing",
    Lifecycle.READY: "completed",
    Lifecycle.FAILED: "draft",
    Lifecycle.ARCHIVED: "archived",
}


def lifecycle_to_legacy_status(lc: Lifecycle) -> str:
    return _LIFECYCLE_TO_LEGACY_STATUS[lc]


# ════════════════════════════════════════════════════════════════════════════
# Error codes — what the service raises, what the route returns
# ════════════════════════════════════════════════════════════════════════════

class StorySessionErrorCode(str, enum.Enum):
    DRAFT_NOT_FOUND = "DRAFT_NOT_FOUND"
    NOT_OWNED = "NOT_OWNED"
    STALE_WRITE = "STALE_WRITE"                       # version mismatch
    ILLEGAL_TRANSITION = "ILLEGAL_TRANSITION"
    SCHEMA_UNSUPPORTED = "DRAFT_SCHEMA_UNSUPPORTED"
    INVALID_PATCH = "INVALID_PATCH"


class StorySessionError(Exception):
    """Domain-specific error. The route layer maps these to HTTP envelopes
    with `code`, `message`, `request_id`, `retryable`. Never construct an
    HTTPException directly inside the service — keep transport concerns out
    of the domain core."""

    def __init__(
        self,
        code: StorySessionErrorCode,
        message: str,
        *,
        status_code: int = 409,
        retryable: bool = False,
        extra: Optional[Dict[str, Any]] = None,
    ):
        super().__init__(message)
        self.code = code
        self.message = message
        self.status_code = status_code
        self.retryable = retryable
        self.extra = extra or {}


# ════════════════════════════════════════════════════════════════════════════
# StorySessionState — the canonical data model
# ════════════════════════════════════════════════════════════════════════════

CURRENT_SCHEMA_VERSION: int = 1


class StorySessionState(BaseModel):
    """The canonical session document, in memory.

    Immutability
    ------------
    The model is configured `frozen=True`. To produce a derived state you
    MUST go through `patched(**fields)` — it returns a new instance with
    the version pre-incremented. This is what the service does on every
    accepted write.
    """

    model_config = ConfigDict(frozen=True, extra="ignore")

    # ── identity (immutable across the lifetime of the draft) ──
    draft_id: str = Field(..., description="canonical draft uuid hex; never reused")
    user_id: str = Field(..., description="owning user id (stringified)")
    schema_version: int = Field(default=CURRENT_SCHEMA_VERSION)

    # ── version for optimistic locking ──
    version: int = Field(default=0, ge=0)

    # ── lifecycle ──
    lifecycle: Lifecycle = Field(default=Lifecycle.IDLE)

    # ── domain fields ──
    title: str = Field(default="")
    story_text: str = Field(default="")
    animation_style: Optional[str] = None
    age_group: Optional[str] = None
    voice_preset: Optional[str] = None

    # ── attached pipeline job (only set during GENERATING/READY/FAILED) ──
    attached_job_id: Optional[str] = None

    # ── timestamps ──
    created_at: str
    updated_at: str
    archived_at: Optional[str] = None

    # ─────────────────────────── derived ────────────────────────────
    @property
    def legacy_status(self) -> str:
        return lifecycle_to_legacy_status(self.lifecycle)

    @property
    def is_terminal(self) -> bool:
        return self.lifecycle == Lifecycle.ARCHIVED

    # ─────────────────────────── builders ───────────────────────────
    def patched(
        self,
        *,
        next_lifecycle: Optional[Lifecycle] = None,
        **fields: Any,
    ) -> "StorySessionState":
        """Return a NEW state with `fields` applied and `version` bumped by 1.

        If `next_lifecycle` is provided, it is validated against the legal
        transition table and stored. Re-asserting the same lifecycle is a
        no-op (idempotent retry support).

        Raises `StorySessionError(ILLEGAL_TRANSITION)` for forbidden moves.
        """
        if next_lifecycle is not None and not is_legal_transition(self.lifecycle, next_lifecycle):
            raise StorySessionError(
                StorySessionErrorCode.ILLEGAL_TRANSITION,
                f"Cannot transition from {self.lifecycle.value} to {next_lifecycle.value}",
                status_code=409,
                retryable=False,
                extra={
                    "from": self.lifecycle.value,
                    "to": next_lifecycle.value,
                    "allowed_next": sorted(s.value for s in legal_next_states(self.lifecycle)),
                },
            )

        # Compute new lifecycle + archived_at side effect cleanly.
        new_lifecycle = next_lifecycle if next_lifecycle is not None else self.lifecycle
        now_iso = datetime.now(timezone.utc).isoformat()
        archived_at = self.archived_at
        if new_lifecycle == Lifecycle.ARCHIVED and archived_at is None:
            archived_at = now_iso

        # Build the new field dict. We deliberately use model_dump + override
        # so unknown future fields (added by `extra='ignore'` rejection)
        # don't leak through.
        data = self.model_dump()
        data.update(fields)
        data["lifecycle"] = new_lifecycle
        data["version"] = self.version + 1
        data["updated_at"] = now_iso
        data["archived_at"] = archived_at
        return StorySessionState(**data)

    # ─────────────────────────── serialization ───────────────────────────
    def to_mongo(self) -> Dict[str, Any]:
        """Return a plain dict suitable for `$set` in MongoDB.

        Always emits both the new `lifecycle` field AND the legacy `status`
        field so older endpoints (and admin tools) keep working. Never emits
        `_id` — Mongo ObjectId remains untouched.
        """
        return {
            "draft_id": self.draft_id,
            "user_id": self.user_id,
            "schema_version": self.schema_version,
            "version": self.version,
            "lifecycle": self.lifecycle.value,
            "status": lifecycle_to_legacy_status(self.lifecycle),  # legacy
            "title": self.title,
            "story_text": self.story_text,
            "animation_style": self.animation_style,
            "age_group": self.age_group,
            "voice_preset": self.voice_preset,
            "attached_job_id": self.attached_job_id,
            "created_at": self.created_at,
            "updated_at": self.updated_at,
            "archived_at": self.archived_at,
        }

    @classmethod
    def from_mongo(cls, doc: Dict[str, Any]) -> "StorySessionState":
        """Build a canonical state from a Mongo document. Backfills missing
        fields with safe defaults so pre-Session-2 docs hydrate without
        requiring a batch migration.

        Raises `StorySessionError(SCHEMA_UNSUPPORTED)` if `schema_version`
        is from the future.
        """
        schema = int(doc.get("schema_version", CURRENT_SCHEMA_VERSION))
        if schema > CURRENT_SCHEMA_VERSION:
            raise StorySessionError(
                StorySessionErrorCode.SCHEMA_UNSUPPORTED,
                "We found an issue restoring this draft. Recover safe content or start fresh.",
                status_code=409,
                retryable=False,
                extra={
                    "schema_version": schema,
                    "supported_version": CURRENT_SCHEMA_VERSION,
                },
            )

        # Lifecycle backfill from legacy `status` if missing.
        lc_raw = doc.get("lifecycle")
        if lc_raw is None:
            legacy = doc.get("status", "draft")
            lc = _LEGACY_STATUS_TO_LIFECYCLE.get(legacy, Lifecycle.IDLE)
            # If the doc has content, upgrade IDLE → EDITING for accuracy.
            if lc == Lifecycle.IDLE and (
                doc.get("title") or doc.get("story_text")
            ):
                lc = Lifecycle.EDITING
        else:
            try:
                lc = Lifecycle(lc_raw)
            except ValueError:
                lc = Lifecycle.EDITING  # safe fallback for corrupt rows

        now_iso = datetime.now(timezone.utc).isoformat()
        return cls(
            draft_id=str(doc.get("draft_id") or doc.get("_id") or ""),
            user_id=str(doc.get("user_id", "")),
            schema_version=schema,
            version=int(doc.get("version", 0)),
            lifecycle=lc,
            title=doc.get("title", "") or "",
            story_text=doc.get("story_text", "") or "",
            animation_style=doc.get("animation_style"),
            age_group=doc.get("age_group"),
            voice_preset=doc.get("voice_preset"),
            attached_job_id=doc.get("attached_job_id"),
            created_at=doc.get("created_at") or now_iso,
            updated_at=doc.get("updated_at") or now_iso,
            archived_at=doc.get("archived_at"),
        )


_LEGACY_STATUS_TO_LIFECYCLE: Dict[str, Lifecycle] = {
    "draft": Lifecycle.IDLE,
    "processing": Lifecycle.GENERATING,
    "completed": Lifecycle.READY,
    "archived": Lifecycle.ARCHIVED,
}


# ════════════════════════════════════════════════════════════════════════════
# Patch DTO — what the service accepts on a write
# ════════════════════════════════════════════════════════════════════════════

class StorySessionPatch(BaseModel):
    """Domain fields a client may patch on EDITING/AUTOSAVING transitions.

    Only the whitelisted fields below are accepted. Anything else is dropped
    silently. The service uses this to enforce least-privilege writes.
    """
    model_config = ConfigDict(extra="forbid")

    title: Optional[str] = None
    story_text: Optional[str] = None
    animation_style: Optional[str] = None
    age_group: Optional[str] = None
    voice_preset: Optional[str] = None

    def non_null_fields(self) -> Dict[str, Any]:
        return {k: v for k, v in self.model_dump().items() if v is not None}


__all__ = [
    "Lifecycle",
    "is_legal_transition",
    "legal_next_states",
    "lifecycle_to_legacy_status",
    "StorySessionState",
    "StorySessionPatch",
    "StorySessionError",
    "StorySessionErrorCode",
    "CURRENT_SCHEMA_VERSION",
]

File: frontend/src/state/storySession.js
/**
 * Story Session State — Frontend canonical model (2026-05-17, Session 2)
 * ======================================================================
 *
 * This module mirrors the backend `StorySessionState` exactly. It is the
 * SINGLE place that knows the lifecycle graph for the frontend.
 *
 * Design pillars (founder-mandated):
 *   1. ONE canonical state object — no localStorage / split useState games
 *   2. Explicit lifecycle transitions enforced in the reducer
 *   3. Illegal transitions are dev-time errors (assertion) AND no-ops in prod
 *   4. Version is a number — server is authoritative
 *   5. Stale write rejection — when the server returns STALE_WRITE the
 *      reducer takes the server's current_version + payload and re-syncs
 *   6. Deterministic hydration — `hydrateFromServer(state)` is idempotent;
 *      replacing state never causes flicker because every selector is pure
 *   7. Single source of truth — components subscribe via `useStorySession`
 *   8. Immutable updates — every reducer action returns a NEW state object
 *   9. Strict ownership — `(draftId, userId)` baked into the state shape
 *  10. Regression coverage first-class — see __tests__/storySession.test.js
 *
 * Scope discipline
 * ----------------
 * No routes, no fetch, no React imports. Pure data. The fetch layer lives
 * in `storySessionClient.js`; the React binding lives in
 * `useStorySession.js`. This split keeps the reducer cheap to test and
 * reason about.
 */

// ─── Lifecycle constants ─────────────────────────────────────────────────
export const Lifecycle = Object.freeze({
  IDLE: 'IDLE',
  EDITING: 'EDITING',
  AUTOSAVING: 'AUTOSAVING',
  READY_TO_GENERATE: 'READY_TO_GENERATE',
  GENERATING: 'GENERATING',
  READY: 'READY',
  FAILED: 'FAILED',
  ARCHIVED: 'ARCHIVED',
});

// Mirror of backend/models/story_session.py `_LEGAL_TRANSITIONS`.
// Any change here MUST be matched on the backend, and vice-versa.
const LEGAL_TRANSITIONS = Object.freeze({
  IDLE:               new Set(['EDITING', 'ARCHIVED']),
  EDITING:            new Set(['AUTOSAVING', 'READY_TO_GENERATE', 'ARCHIVED']),
  AUTOSAVING:         new Set(['EDITING', 'READY_TO_GENERATE', 'ARCHIVED']),
  READY_TO_GENERATE:  new Set(['EDITING', 'GENERATING', 'ARCHIVED']),
  GENERATING:         new Set(['READY', 'FAILED']),
  READY:              new Set(['ARCHIVED']),
  FAILED:             new Set(['EDITING', 'ARCHIVED']),
  ARCHIVED:           new Set(),
});

export function isLegalTransition(prev, next) {
  if (prev === next) return true;        // idempotent retries
  const allowed = LEGAL_TRANSITIONS[prev];
  return allowed ? allowed.has(next) : false;
}

export function legalNextStates(prev) {
  const allowed = LEGAL_TRANSITIONS[prev];
  return allowed ? Array.from(allowed).sort() : [];
}

// ─── Error codes (match backend StorySessionErrorCode) ──────────────────
export const ErrorCode = Object.freeze({
  DRAFT_NOT_FOUND: 'DRAFT_NOT_FOUND',
  NOT_OWNED: 'NOT_OWNED',
  STALE_WRITE: 'STALE_WRITE',
  ILLEGAL_TRANSITION: 'ILLEGAL_TRANSITION',
  SCHEMA_UNSUPPORTED: 'DRAFT_SCHEMA_UNSUPPORTED',
  INVALID_PATCH: 'INVALID_PATCH',
  // P0 2026-05-30 — DRAFT_ALREADY_ACTIVE is a RECOVERABLE state, not a
  // failure. Backend returns 409 + `active_draft_id` so the client can
  // adopt the existing draft. Surfacing this as a generic error toast
  // (the "Couldn't start a new draft. Ref: ..." bug) strands the user.
  // Canonical recovery path: useStorySessionAutosave adopts the
  // existing draft on this code. Pinned by audit
  // test_draft_already_active_recovery_2026_05.py.
  DRAFT_ALREADY_ACTIVE: 'DRAFT_ALREADY_ACTIVE',
});

// ─── Canonical state shape ───────────────────────────────────────────────
export const initialStorySessionState = Object.freeze({
  // identity (immutable across draft lifetime)
  draftId: null,
  userId: null,
  schemaVersion: 1,

  // version + lifecycle
  version: 0,
  lifecycle: Lifecycle.IDLE,

  // domain fields
  title: '',
  storyText: '',
  animationStyle: null,
  ageGroup: null,
  voicePreset: null,

  // attached pipeline job
  attachedJobId: null,

  // timestamps
  createdAt: null,
  updatedAt: null,
  archivedAt: null,

  // client-side metadata (NOT sent to server)
  meta: Object.freeze({
    // last server-correlated request id — surfaces in error UX
    lastRequestId: null,
    // last error (only one in-flight at a time) — { code, message, retryable }
    lastError: null,
    // whether a save is currently in flight
    saving: false,
  }),
});

// ─── Action types ────────────────────────────────────────────────────────
export const Actions = Object.freeze({
  // Server-driven hydration. Single source of truth for incoming state.
  HYDRATE: 'storySession/HYDRATE',

  // Local edit — used by uncontrolled inputs before autosave runs.
  // Does NOT bump version (version is server-authoritative).
  LOCAL_EDIT: 'storySession/LOCAL_EDIT',

  // Save lifecycle markers.
  SAVE_PENDING: 'storySession/SAVE_PENDING',
  SAVE_OK: 'storySession/SAVE_OK',
  SAVE_FAILED: 'storySession/SAVE_FAILED',

  // Stale-write recovery — server returned the canonical state, take it.
  RESYNC: 'storySession/RESYNC',

  // Hard reset (Start Fresh).
  RESET: 'storySession/RESET',

  // Error envelope passthrough (no state change, just metadata).
  ERROR: 'storySession/ERROR',
});

// ─── Pure reducer ────────────────────────────────────────────────────────
// Always returns a NEW object. Never mutates. Frozen-input safe.
export function storySessionReducer(state, action) {
  const s = state || initialStorySessionState;
  switch (action.type) {
    case Actions.HYDRATE:
    case Actions.RESYNC: {
      // Server is the authority. Replace identity + version + lifecycle +
      // domain fields. Preserve client-side `meta` unless `meta` was
      // explicitly provided.
      const payload = action.payload || {};
      return {
        ...s,
        draftId: payload.draftId ?? s.draftId,
        userId: payload.userId ?? s.userId,
        schemaVersion: payload.schemaVersion ?? s.schemaVersion,
        version: payload.version ?? s.version,
        lifecycle: payload.lifecycle ?? s.lifecycle,
        title: payload.title ?? '',
        storyText: payload.storyText ?? '',
        animationStyle: payload.animationStyle ?? null,
        ageGroup: payload.ageGroup ?? null,
        voicePreset: payload.voicePreset ?? null,
        attachedJobId: payload.attachedJobId ?? null,
        createdAt: payload.createdAt ?? s.createdAt,
        updatedAt: payload.updatedAt ?? s.updatedAt,
        archivedAt: payload.archivedAt ?? null,
        meta: {
          ...s.meta,
          saving: false,
          lastError: null,
          lastRequestId: payload.requestId ?? s.meta.lastRequestId,
        },
      };
    }

    case Actions.LOCAL_EDIT: {
      // Lifecycle guard: editing is only legal from IDLE/EDITING/AUTOSAVING/
      // READY_TO_GENERATE/FAILED. GENERATING/READY/ARCHIVED reject edits.
      if (
        s.lifecycle === Lifecycle.GENERATING ||
        s.lifecycle === Lifecycle.READY ||
        s.lifecycle === Lifecycle.ARCHIVED
      ) {
        // dev-time loud, prod-time silent (last-write-wins is not legal here)
        if (process.env.NODE_ENV !== 'production') {
          // eslint-disable-next-line no-console
          console.warn(
            `[storySession] LOCAL_EDIT rejected in ${s.lifecycle} state`
          );
        }
        return s;
      }
      const patch = action.payload || {};
      const nextLifecycle =
        s.lifecycle === Lifecycle.IDLE ? Lifecycle.EDITING : s.lifecycle;
      return {
        ...s,
        title: patch.title ?? s.title,
        storyText: patch.storyText ?? s.storyText,
        animationStyle: patch.animationStyle ?? s.animationStyle,
        ageGroup: patch.ageGroup ?? s.ageGroup,
        voicePreset: patch.voicePreset ?? s.voicePreset,
        lifecycle: nextLifecycle,
      };
    }

    case Actions.SAVE_PENDING:
      return { ...s, meta: { ...s.meta, saving: true, lastError: null } };

    case Actions.SAVE_OK: {
      // Server returned canonical state — same as HYDRATE but flagged.
      const payload = action.payload || {};
      return {
        ...s,
        version: payload.version ?? s.version,
        lifecycle: payload.lifecycle ?? s.lifecycle,
        updatedAt: payload.updatedAt ?? s.updatedAt,
        attachedJobId: payload.attachedJobId ?? s.attachedJobId,
        archivedAt: payload.archivedAt ?? s.archivedAt,
        meta: {
          ...s.meta,
          saving: false,
          lastError: null,
          lastRequestId: payload.requestId ?? s.meta.lastRequestId,
        },
      };
    }

    case Actions.SAVE_FAILED: {
      const err = action.payload || {};
      return {
        ...s,
        meta: {
          ...s.meta,
          saving: false,
          lastError: {
            code: err.code || 'UNKNOWN',
            message: err.message || 'Save failed.',
            retryable: !!err.retryable,
          },
          lastRequestId: err.requestId ?? s.meta.lastRequestId,
        },
      };
    }

    case Actions.RESET:
      return { ...initialStorySessionState };

    case Actions.ERROR: {
      const err = action.payload || {};
      return {
        ...s,
        meta: {
          ...s.meta,
          lastError: {
            code: err.code || 'UNKNOWN',
            message: err.message || '',
            retryable: !!err.retryable,
          },
          lastRequestId: err.requestId ?? s.meta.lastRequestId,
        },
      };
    }

    default:
      return s;
  }
}

// ─── Action creators ─────────────────────────────────────────────────────
export const actions = {
  hydrate: (payload) => ({ type: Actions.HYDRATE, payload }),
  resync: (payload) => ({ type: Actions.RESYNC, payload }),
  localEdit: (patch) => ({ type: Actions.LOCAL_EDIT, payload: patch }),
  savePending: () => ({ type: Actions.SAVE_PENDING }),
  saveOk: (payload) => ({ type: Actions.SAVE_OK, payload }),
  saveFailed: (err) => ({ type: Actions.SAVE_FAILED, payload: err }),
  reset: () => ({ type: Actions.RESET }),
  error: (err) => ({ type: Actions.ERROR, payload: err }),
};

// ─── Selectors (pure) ────────────────────────────────────────────────────
export const selectors = {
  isSaving: (s) => !!s?.meta?.saving,
  isArchived: (s) => s?.lifecycle === Lifecycle.ARCHIVED,
  isGenerating: (s) => s?.lifecycle === Lifecycle.GENERATING,
  canEdit: (s) =>
    s &&
    s.lifecycle !== Lifecycle.GENERATING &&
    s.lifecycle !== Lifecycle.READY &&
    s.lifecycle !== Lifecycle.ARCHIVED,
  canGenerate: (s) =>
    !!s &&
    (s.lifecycle === Lifecycle.EDITING ||
      s.lifecycle === Lifecycle.AUTOSAVING ||
      s.lifecycle === Lifecycle.READY_TO_GENERATE ||
      s.lifecycle === Lifecycle.FAILED) &&
    (!!s.title?.trim() || !!s.storyText?.trim()),
  lastError: (s) => s?.meta?.lastError || null,
  lastRequestId: (s) => s?.meta?.lastRequestId || null,
  legalNext: (s) => legalNextStates(s?.lifecycle || Lifecycle.IDLE),
};

// ─── Wire-format mapping (server <-> client field names) ─────────────────
// The backend uses snake_case (draft_id, story_text, attached_job_id, etc.)
// The frontend uses camelCase. This is the SINGLE place that knows the map.
export function fromServerState(serverState, requestId = null) {
  if (!serverState) return null;
  return {
    draftId: serverState.draft_id,
    schemaVersion: serverState.schema_version,
    version: serverState.version,
    lifecycle: serverState.lifecycle,
    title: serverState.title || '',
    storyText: serverState.story_text || '',
    animationStyle: serverState.animation_style || null,
    ageGroup: serverState.age_group || null,
    voicePreset: serverState.voice_preset || null,
    attachedJobId: serverState.attached_job_id || null,
    createdAt: serverState.created_at || null,
    updatedAt: serverState.updated_at || null,
    archivedAt: serverState.archived_at || null,
    allowedNext: serverState.allowed_next || [],
    requestId,
  };
}
